tools/fpgabist/bist/bist_afu.c

```diff
@@ -0,0 +1,387 @@
+	ON_ERR_GOTO(res, out_free_input, "allocating output buffer");
+
+	printf("Running Test\n");
+
+	/* Initialize buffers */
+	memset((void *)dsm_ptr,    0,    LPBK1_DSM_SIZE);
+	memset((void *)input_ptr,  0xAF, LPBK1_BUFFER_SIZE);
+	memset((void *)output_ptr, 0xBE, LPBK1_BUFFER_SIZE);
+
+	cache_line *cl_ptr = (cache_line *)input_ptr;
+	for (uint32_t i = 0; i < LPBK1_BUFFER_SIZE / CL(1); ++i) {
+		cl_ptr[i].uint[15] = i+1; /* set the last uint in every cacheline */
+	}
+
+	/* Reset accelerator */
+	res = fpgaReset(accelerator_handle);
+	ON_ERR_GOTO(res, out_free_output, "resetting accelerator");
+
+	/* Program DMA addresses */
+	uint64_t iova;
+	res = fpgaGetIOAddress(accelerator_handle, dsm_wsid, &iova);
+	ON_ERR_GOTO(res, out_free_output, "getting DSM IOVA");
+
+	res = fpgaWriteMMIO64(accelerator_handle, 0, CSR_AFU_DSM_BASEL, iova);
+	ON_ERR_GOTO(res, out_free_output, "writing CSR_AFU_DSM_BASEL");
+
+	res = fpgaWriteMMIO32(accelerator_handle, 0, CSR_CTL, 0);
+	ON_ERR_GOTO(res, out_free_output, "writing CSR_CFG");
+	res = fpgaWriteMMIO32(accelerator_handle, 0, CSR_CTL, 1);
+	ON_ERR_GOTO(res, out_free_output, "writing CSR_CFG");
+
+	res = fpgaGetIOAddress(accelerator_handle, input_wsid, &iova);
+	ON_ERR_GOTO(res, out_free_output, "getting input IOVA");
+	res = fpgaWriteMMIO64(accelerator_handle, 0, CSR_SRC_ADDR, CACHELINE_ALIGNED_ADDR(iova));
+	ON_ERR_GOTO(res, out_free_output, "writing CSR_SRC_ADDR");
+
+	res = fpgaGetIOAddress(accelerator_handle, output_wsid, &iova);
+	ON_ERR_GOTO(res, out_free_output, "getting output IOVA");
+	res = fpgaWriteMMIO64(accelerator_handle, 0, CSR_DST_ADDR, CACHELINE_ALIGNED_ADDR(iova));
+	ON_ERR_GOTO(res, out_free_output, "writing CSR_DST_ADDR");
+	//fpgaProgramBufferAddressAndLength(accelerator_handle, dsm_wsid, 0, LPBK1_DSM_SIZE,
+	//				   CSR_AFU_DSM_BASEL);
+	//fpgaProgramBufferAddressAndLength(accelerator_handle, input_wsid, 0, LPBK1_BUFFER_SIZE,
+	//				   CSR_SRC_ADDR);
+	//fpgaProgramBufferAddressAndLength(accelerator_handle, output_wsid, 0, LPBK1_BUFFER_SIZE,
+	//				   CSR_DST_ADDR);
+
+	res = fpgaWriteMMIO32(accelerator_handle, 0, CSR_NUM_LINES, LPBK1_BUFFER_SIZE / CL(1));
+	ON_ERR_GOTO(res, out_free_output, "writing CSR_NUM_LINES");
+	res = fpgaWriteMMIO32(accelerator_handle, 0, CSR_CFG, 0x42000);
+	ON_ERR_GOTO(res, out_free_output, "writing CSR_CFG");
+
+	status_ptr = dsm_ptr + DSM_STATUS_TEST_COMPLETE/8;
+
+	/* Start the test */
+	res = fpgaWriteMMIO32(accelerator_handle, 0, CSR_CTL, 3);
+	ON_ERR_GOTO(res, out_free_output, "writing CSR_CFG");
+
+	/* Wait for test completion */
+	while (0 == ((*status_ptr) & 0x1)) {
+		usleep(100);
+	}
+
+
+/************* Begin BIST *************/
+        /* Perform BIST Check */
+	printf("Running BIST Test\n");
+        uint64_t data = 0;
+        double count = 0;
+        unsigned int bist_mask = ENABLE_DDRA_BIST;
+        unsigned int ddra_bist_result = 0;
+
+	res = fpgaWriteMMIO32(accelerator_handle, 0, DDR_BIST_CTRL_ADDR, bist_mask);
```

Comment:
Why don't you create separate functions performing check?

Reply:
Right now with the GOTO's abstaction of C code is increasingly complicated.  There will need to be some future effort to address these concerns in future applications.